File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        ## space complexity:O(1)
        ## time complexity:O(n^2)

        m, n = len(matrix), len(matrix[0])

        for i in range(0, m//2):
            matrix[~i], matrix[i] = matrix[i], matrix[~i]

        for i in range(m):
            for j in range(i):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        
        int m = matrix.length-1;
        int n = matrix[0].length-1;

        for(int i = 0; i <= m / 2;  i++){
            int[] temp = matrix[i];
            matrix[i] = matrix[m-i];
            matrix[m-i] = temp;
        }

        for(int i = 0; i <= m; i++){
            for(int j = 0; j < i; j++){
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
    }
}